### PR TITLE
Normalize TypedArray, ArrayBuffer and DataView constructors 

### PR DESCRIPTION
**Pull request by leobalter**
Ref #265

On these changes, the offset arguments have a consistent check to throw for values < 0, and length will pass through the ToLength.

### CODE REVIEW

**Comment by anba** [comment]
I'd avoid adding explicit tests to check whether or not an optional parameter is present, if normal type conversions through ToInteger/ToLength are sufficient. And [I still think](https://github.com/tc39/ecma262/issues/265#issuecomment-165837760) it's important to find out why the current parameter validation steps were added. And if the validation steps in the DataView constructor are changed, we may also want to apply [similar changes to GetViewValue/SetViewValue](https://bugs.ecmascript.org/show_bug.cgi?id=4516#c2).

**Comment by leobalter** [comment]
> I'd avoid adding explicit tests to check whether or not an optional parameter is present, if normal type conversions through ToInteger/ToLength are sufficient.

I tried to follow other parts handling arguments that are not present, even when they get the undefined value.

> And I still think it's important to find out why the current parameter validation steps were added.

I'm on the same boat.

> we may also want to apply similar changes to GetViewValue/SetViewValue.

+1, I'll include these

**Comment by leobalter** [comment]
@anba the new commits are addressing some extras, including the _TypedArray_ (length) constructor.

It still does not answer the parameter validation question. I'll check if I find anything.

**Comment by anba** [comment]
> I tried to follow other parts handling arguments that are not present, even when they get the undefined value.

I guess it depends on the built-ins, for example Array.prototype.copyWithin, Array.prototype.fill, Array.prototype.includes, Array.prototype.indexOf and %TypedArray%.prototype.set all use implicit conversion instead of explicit checks for non-present parameters.

File: spec.html [hunk 1]
```diff
@@ -33172,19 +33166,19 @@
...
-          1. Let _offset_ be ToInteger(_numberOffset_).
-          1. If _numberOffset_ &ne; _offset_ or _offset_ &lt; 0, throw a *RangeError* exception.
+          1. Let _offset_ be ? ToInteger(_byteOffset_).
+          1. If _offset_ &lt; 0, throw a *RangeError* exception.
+          1. If _offset_ is -0, let _offset_ be +0.
           1. If IsDetachedBuffer(_buffer_) is *true*, throw a *TypeError* exception.
           1. Let _bufferByteLength_ be the value of _buffer_'s [[ArrayBufferByteLength]] internal slot.
           1. If _offset_ &gt; _bufferByteLength_, throw a *RangeError* exception.
-          1. If _byteLength_ is *undefined*, then
+          1. If _byteLength_ is either not present or *undefined*, then
```
Comment (leobalter) [inline spec.html]: On the last commit I removed the "not present" steps, but I wanted to keep this one as it still bugs me here.

Comment (leobalter) [inline spec.html]: @anba ^^

File: spec.html [hunk 2]
```diff
@@ -33172,19 +33166,19 @@
 
       <!-- es6num="24.2.2.1" -->
       <emu-clause id="sec-dataview-buffer-byteoffset-bytelength">
-        <h1>DataView (_buffer_, _byteOffset_, _byteLength_ )</h1>
-        <p>`DataView` called with arguments _buffer_, _byteOffset_, and _byteLength_ performs the following steps:</p>
+        <h1>DataView (_buffer_ [ , _byteOffset_, _byteLength_ ] )</h1>
+        <p>When the `DataView` is called with at least one argument _buffer_, the following steps are taken:</p>
```
Comment (annevk) [inline spec.html]: function*

File: spec.html [hunk 3]
```diff
@@ -32989,13 +32987,11 @@
 
       <!-- es6num="24.1.2.1" -->
       <emu-clause id="sec-arraybuffer-length">
-        <h1>ArrayBuffer ( _length_ )</h1>
-        <p>ArrayBuffer called with argument _length_ performs the following steps:</p>
+        <h1>ArrayBuffer ( [ _length_ ] )</h1>
```
Comment (anba) [inline spec.html]: Brackets need to be removed to avoid changing `ArrayBuffer.length`.

File: spec.html [hunk 4]
```diff
@@ -33172,19 +33166,19 @@
 
       <!-- es6num="24.2.2.1" -->
       <emu-clause id="sec-dataview-buffer-byteoffset-bytelength">
-        <h1>DataView (_buffer_, _byteOffset_, _byteLength_ )</h1>
-        <p>`DataView` called with arguments _buffer_, _byteOffset_, and _byteLength_ performs the following steps:</p>
+        <h1>DataView (_buffer_ [ , _byteOffset_, _byteLength_ ] )</h1>
```
Comment (anba) [inline spec.html]: Brackets need to be removed to avoid changing `DataView.length`.

File: spec.html [hunk 5]
```diff
@@ -33172,19 +33166,19 @@
...
         <emu-alg>
           1. If NewTarget is *undefined*, throw a *TypeError* exception.
           1. If Type(_buffer_) is not Object, throw a *TypeError* exception.
           1. If _buffer_ does not have an [[ArrayBufferData]] internal slot, throw a *TypeError* exception.
-          1. Let _numberOffset_ be ? ToNumber(_byteOffset_).
-          1. Let _offset_ be ToInteger(_numberOffset_).
-          1. If _numberOffset_ &ne; _offset_ or _offset_ &lt; 0, throw a *RangeError* exception.
+          1. Let _offset_ be ? ToInteger(_byteOffset_).
+          1. If _offset_ &lt; 0, throw a *RangeError* exception.
+          1. If _offset_ is -0, let _offset_ be +0.
```
Comment (anba) [inline spec.html]: :+1: Missed that one when I fixed the similar issue for the typed array constructors.

File: spec.html (hunk 3 above)
Comment (ljharb) [inline spec.html]: It's an optional argument tho - `ArrayBuffer().byteLength === 0`

Comment (anba) [inline spec.html]: But the `length` property should describe the typical number of arguments, and ArrayBuffers are seldom called with no arguments. [19.2.4.1](http://tc39.github.io/ecma262/#sec-function-instances-length):
> The value of the length property is an integer that indicates the typical number of arguments expected by the function.

Comment (leobalter) [inline spec.html]: I agree they are seldom called with no arguments, but it's more clear to have them being explicitly optional, otherwise I would seem like an error to call `new ArrayBuffer()`.

File: spec.html (hunk 4 above)
Comment (leobalter) [inline spec.html]: for the same clarity reason I mentioned on the ArrayBuffer constructor, I believe the brackets are good to tell they are optional.

File: spec.html (hunk 3 above)
Comment (ljharb) [inline spec.html]: Function lengths aren't supposed to be about the "typical" number of arguments, they're supposed to be about the number of required arguments. 19.2.4.1 must have simply been overlooked as part of https://github.com/tc39/ecma262/pull/266

Comment (anba) [inline spec.html]: No, that's not true. Unless you're also advocating to change the length of Object (length is 1, required 0), Function (length is 1, required 0), Array (length is 1, required 0), String (length is 1, required 0), Number (length is 1, required 0), Boolean (length is 1, required 0), Date (length is 7, required 0), etc. etc. :-/

Comment (leobalter) [inline spec.html]: Object has an optional argument and an extraordinary length value. https://tc39.github.io/ecma262/#sec-object-value

We may follow this and define an extraordinary length value for ArrayBuffer.

String, Number and Boolean constructors could be revisited to follow Object and (many other objects) declaring the optional argument.

TBQH, I believe classic constructors as Array, Object and especially Date inherits the length value from the previous definition where it accounted all the possible arguments. The length value is not critical for average use of JS, but this could be discussed on another topic, and eventually fixed with a more consistent definition.

File: spec.html (hunk 2 above)
Comment (leobalter) [inline spec.html]: thanks! 174e72c

File: spec.html (hunk 3 above)
Comment (ljharb) [inline spec.html]: There are a number of legacy cases (most of them constructors or Date methods), this is true, but in the general case we should strive to make the function's length match the number of required arguments.

**Comment by leobalter** [comment]
@littledan I tried to address #468 on 0157b3f. 

I reused this same PR as it already touches the same steps. I can split it to a new PR if it works better.

Note that for consistency this PR also removes the RangeError on `SameValueZero(numberLength, elementLength)`. This check is necessary and we might need to use to ToNumber instead of ToInteger if we want to throw for float values as well. I'll do it if necessary.

**Comment by littledan** [comment]
These changes look good to me, and have the effect of making many things more web-compatible, as per #265. I am not sure, but it still feels like the resulting spec has potential web compatibility concerns, though I'm not nearly as worried. @allenwb What do you think?

**Comment by leobalter** [comment]
@littledan 361930a

**Comment by littledan** [comment]
@leobalter the fix-up looks good, modulo web compat concerns. Looks like a faithful implementation of @allenwb's suggestion, and I'm in favor of the spec change given that the compat issue is preexisting.

File: spec.html [hunk 6]
```diff
@@ -32981,9 +32983,10 @@
         <p>When the `ArrayBuffer` function is called with optional argument _length_, the following steps are taken:</p>
         <emu-alg>
           1. If NewTarget is *undefined*, throw a *TypeError* exception.
-          1. Let _integerLength_ be ToInteger(_length_).
-          1. If _integerLength_ &lt; 0, throw a *RangeError* exception.
-          1. Let _byteLength_ be ? ToLength(_integerLength_).
+          1. Let _numberLength_ be ToNumber(_length_).
```
Comment (jorendorff) [inline spec.html]: Looks like this is missing a `?`.

File: spec.html [hunk 7]
```diff
@@ -32981,9 +32983,10 @@
...
+          1. If _numberLength_ &lt; 0, throw a *RangeError* exception.
+          1. Let _byteLength_ be ? ToLength(_numberLength_).
+          1. If SameValueZero(_numberLength_, _byteLength_) is *false*, throw a *RangeError* exception.
```
Comment (jorendorff) [inline spec.html]: I think with this change, `new ArrayBuffer()` throws a RangeError, because then numberLength is NaN and byteLength is 0.

**Comment by leobalter** [comment]
I tried to address the last concerns pointed by @jorendorff but I'm afraid I might be bike shedding on the behaviors. I'll try to list them here:

- throw RangeError for non integer or negative lengths
  - [x] `_TypedArray_ ( length )`
  - [x] `_TypedArray_ ( buffer, byteoffset, length )`
  - [x] `ArrayBuffer( length )`
  - [x] `DataView( buffer, byteOffset, byteLength )`
- take non-present or undefined length argument as 0.
  - [x] `_TypedArray_ ( length )`
  - [x] `_TypedArray_ ( buffer, byteoffset, length )`
  - [x] `ArrayBuffer( length )`
  - [x] `DataView( buffer, byteOffset, byteLength )`
- length becomes an optional argument
  - [x] `_TypedArray_ ( buffer, byteoffset, length )`
  - [x] `ArrayBuffer( length )`
  - [x] `DataView( buffer, byteOffset, byteLength )`
- accept string values to represent length, as `"0"`
  - [x] `_TypedArray_ ( length )`
  - [x] `_TypedArray_ ( buffer, byteoffset, length )`
  - [x] `ArrayBuffer( length )`
  - [x] `DataView( buffer, byteOffset, byteLength )`

**Comment by leobalter** [comment]
@littledan @bterlson should I add this to the agenda? It might need consensus.

**Comment by bterlson** [comment]
@leobalter Yeah probably a good idea. I'll add the label.

**Comment by bterlson** [comment]
I believe this was discussed at tc39. @leobalter can you summarize the outcome of the discussion? I imagine we are free to fix #265?

**Comment by leobalter** [comment]
- [ ] I need to provide a table with the current results from different runtimes. 
- [ ] I'll also try to provide a table with the data I've showed at the meeting
- [ ] I need to update the current proposal to cast non-integer numbers as integers

@zloirock, will the integer cast help?

**Comment by leobalter** [comment]
I'll provide more data with the expected results from the specs and expected results after change.

This current data is already helpful detect some opportunity for improvements on my changes.

I used v8 5.1.0, SpiderMonkey c47.0a1, Safari 9.0 and Edge (build 25.x + EdgeHTML 13) to fetch these data. 

## _TypedArray_( _length_ )

Results for `new Int8Array( x ).length`:

|      | 0 | 1.1        | 0.1        | -1         | -0 | undefined | null      | Infinity   | NaN        |
|------|---|------------|------------|------------|----|-----------|-----------|------------|------------|
| v8   | 0 | 1          | 0          | RangeError | 0  | 0         | TypeError | RangeError | 0          |
| sm   | 0 | TypeError  | TypeError  | TypeError  | 0  | TypeError | TypeError | TypeError  | TypeError  |
| jsc* | 0 | TypeError  | TypeError  | RangeError | 0  | TypeError | TypeError | TypeError  | TypeError  |
| edge | 0 | RangeError | RangeError | RangeError | 0  | TypeError | TypeError | RangeError | RangeError |

`*` - used MacOS Safari 9.0

## _TypedArray_( buffer, byteOffset, _length_ )

Results for `new Int8Array( new ArrayBuffer(128), 0, x ).length`:

|      | n/a | 0 | 1.1        | 0.1        | -1         | -0 | undefined | null      | Infinity   | NaN        |
|------|-----|---|------------|------------|------------|----|-----------|-----------|------------|------------|
| v8   | 128 | 0 | 1          | 0          | RangeError | 0  | 128       | 0         | RangeError | 0          |
| sm   | 128 | 0 | 1          | 0          | RangeError | 0  | 0         | 0         | 0          | 0          |
| jsc* | 128 | 0 | 1          | 0          | RangeError | 0  | 0         | 0         | 0          | 0          |
| edge | 128 | 0 | 1          | 0          | RangeError | 0  | 128       | 0         | 0          | 0          |

## ArrayBuffer( _length_ )

Results for `new ArrayBuffer( x ).byteLength`:

|      | n/a | 0 | 1.1        | 0.1        | -1         | -0 | undefined | null      | Infinity   | NaN        |
|------|-----|---|------------|------------|------------|----|-----------|-----------|------------|------------|
| v8   | 0   | 0 | 1          | 0          | RangeError | 0  | 0         | 0         | RangeError | 0          |
| sm   | 0   | 0 | 1          | 0          | RangeError | 0  | 0         | 0         | 0          | 0          |
| jsc* | 0   | 0 | 1          | 0          | Error      | 0  | 0         | 0         | 0          | 0          |
| edge | 0   | 0 | 1          | 0          | RangeError | 0  | 0         | 0         | 0          | 0          |

## DataView( buffer, byteOffset, _byteLength_ )

Results for `new DataView( new ArrayBuffer(128), 0, x ).byteLength`:

|      | n/a | 0 | 1.1        | 0.1        | -1         | -0 | undefined | null      | Infinity   | NaN        |
|------|-----|---|------------|------------|------------|----|-----------|-----------|------------|------------|
| v8   | 128 | 0 | 1          | 0          | 0          | 0  | 128       | 0         | RangeError | 0          |
| sm   | 128 | 0 | 1          | 0          | RangeError | 0  | 128       | 0         | 0          | 0          |
| jsc* | 128 | 0 | 1          | 0          | RangeError | 0  | 0         | 0         | 0          | 0          |
| edge | 128 | 0 | 1          | 0          | 0          | 0  | 128       | 0         | RangeError | 0          |

**Comment by domenic** [comment]
That is super-great investigative work. Looking forward to seeing your proposal given that. It seems exciting that so many people throw errors for bad input, since it means we can probably throw errors for bad input in the spec too.

**Comment by leobalter** [comment]
I've got new up-to-date tables including the expected results from the specs (same on ES6, ES7, and draft) and the expected results with the current PR.

The changes on this PR are now updated to reflect this table.

## _TypedArray_( _length_ )

Results for `new Int8Array( x ).length`:

|       | 0 | 1.1        | 0.1        | -1         | -0 | undefined | null      | Infinity   | NaN        |
|-------|---|------------|------------|------------|----|-----------|-----------|------------|------------|
| v8    | 0 | 1          | 0          | RangeError | 0  | 0         | TypeError | RangeError | 0          |
| sm    | 0 | TypeError  | TypeError  | TypeError  | 0  | TypeError | TypeError | TypeError  | TypeError  |
| jsc*  | 0 | TypeError  | TypeError  | RangeError | 0  | TypeError | TypeError | TypeError  | TypeError  |
| edge  | 0 | RangeError | RangeError | RangeError | 0  | TypeError | TypeError | RangeError | RangeError |
| specs | 0 | RangeError | RangeError | RangeError | 0  | TypeError | 0         | RangeError | RangeError |
| PR    | 0 | 1          | 0          | RangeError | 0  | 0         | 0         | RangeError | 0          |

`*` - used MacOS Safari 9.0

## _TypedArray_( buffer, byteOffset, _length_ )

Results for `new Int8Array( new ArrayBuffer(128), 0, x ).length`:

|       | n/a | 0 | 1.1 | 0.1 | -1         | -0 | undefined | null | Infinity    | NaN |
|-------|-----|---|-----|-----|------------|----|-----------|------|-------------|-----|
| v8    | 128 | 0 | 1   | 0   | RangeError | 0  | 128       | 0    | RangeError  | 0   |
| sm    | 128 | 0 | 1   | 0   | RangeError | 0  | 0         | 0    | 0           | 0   |
| jsc*  | 128 | 0 | 1   | 0   | RangeError | 0  | 0         | 0    | 0           | 0   |
| edge  | 128 | 0 | 1   | 0   | RangeError | 0  | 128       | 0    | 0           | 0   |
| specs | 128 | 0 | 1   | 0   | 0          | 0  | 128       | 0    | 2 ** 53 - 1 | 0   |
| PR    | 128 | 0 | 1   | 0   | RangeError | 0  | 128       | 0    | RangeError  | 0   |

## ArrayBuffer( _length_ )

Results for `new ArrayBuffer( x ).byteLength`:

|       | n/a        | 0          | 1.1        | 0.1        | -1         | -0 | undefined  | null | Infinity   | NaN        |
|-------|------------|------------|------------|------------|------------|----|------------|------|------------|------------|
| v8    | 0          | 0          | 1          | 0          | RangeError | 0  | 0          | 0    | RangeError | 0          |
| sm    | 0          | 0          | 1          | 0          | RangeError | 0  | 0          | 0    | 0          | 0          |
| jsc*  | 0          | 0          | 1          | 0          | Error      | 0  | 0          | 0    | 0          | 0          |
| edge  | 0          | 0          | 1          | 0          | RangeError | 0  | 0          | 0    | 0          | 0          |
| specs | RangeError | RangeError | RangeError | RangeError | RangeError | 0  | RangeError | 0    | RangeError | RangeError |
| PR    | 0          | 0          | 1          | 0          | RangeError | 0  | 0          | 0    | RangeError | 0          |

## DataView( buffer, byteOffset, _byteLength_ )

Results for `new DataView( new ArrayBuffer(128), 0, x ).byteLength`:

|       | n/a | 0 | 1.1        | 0.1        | -1         | -0 | undefined | null      | Infinity    | NaN        |
|-------|-----|---|------------|------------|------------|----|-----------|-----------|-------------|------------|
| v8    | 128 | 0 | 1          | 0          | 0          | 0  | 128       | 0         | RangeError  | 0          |
| sm    | 128 | 0 | 1          | 0          | RangeError | 0  | 128       | 0         | 0           | 0          |
| jsc*  | 128 | 0 | 1          | 0          | RangeError | 0  | 0         | 0         | 0           | 0          |
| edge  | 128 | 0 | 1          | 0          | 0          | 0  | 128       | 0         | RangeError  | 0          |
| draft | 128 | 0 | 1          | 0          | 0          | 0  | 128       | 0         | 2 ** 53 - 1 | 0          |
| PR    | 128 | 0 | 1          | 0          | RangeError | 0  | 128       | 0         | RangeError  | 0          |

**Comment by littledan** [comment]
Excellent data. I'm happy with these semantics; they seem to be good with respect to both consistency and probable web compatibility (at least from an intersection semantics point of view). It's nice that it's shown here that new exceptions can be thrown in some cases where the ES2017 draft spec currently doesn't throw.

@allenwb Have you found more information which you mentioned about previous motivation to not throw errors in some of these cases? Or are you skeptical of some of the checks which are removed here?

**Comment by leobalter** [comment]
I've also browsed at the Khronos spec, it wasn't helpful to improve my data: https://www.khronos.org/registry/typedarray/specs/latest/ Maybe I should seek for another info from Khronos to find some historical background.

**Comment by allenwb** [comment]
@leobalter 
In your table for: new Int8Array( new ArrayBuffer(128), 0, x ).length 
you say the spec. produces 2**53-1 when x is +∞.  My reading of both the ES2015 spec. and the current draft is that RangeError should occur. See [Step 14.c](http://tc39.github.io/ecma262/#sec-typedarray-buffer-byteoffset-length). 
Similar issue for the DateView table. See [Ste3p 11.b](http://tc39.github.io/ecma262/#sec-dataview-buffer-byteoffset-bytelength).
What are you interpreting differently from me?

Regarding the [Khronos spec.](https://www.khronos.org/registry/typedarray/specs/latest/) I think the original author (Ken) just used WebIDL conventions to describe the signatures  To understand them  you have to look at the Khronos signatures and then look at the conversion rules for the WebIDL types used in the [WebIDL spec that was contemporary](https://www.w3.org/TR/WebIDL/) with the Khronos spec. For example, x in the signature applicable to the above table is specified to be "unsigned long" so the [this conversion](https://www.w3.org/TR/WebIDL/#es-unsigned-long) applies to it.

@littledan Regarding "previous motivations", I was thinking about things like the following. But I didn't finde anything that is immediately relevant.  There are also probably closed issues bugs.ecmascript.org that may be relevant.   

```
From: Allen Wirfs-Brock <allenwb@mozilla.com>
Subject: Re: double-checking on typed array indices
Date: February 28, 2014 at 9:34:46 AM PST
To: Luke Wagner <luke@mozilla.com>
Cc: Brendan Eich <brendan@mozilla.com>


On Feb 28, 2014, at 7:36 AM, Luke Wagner wrote:

In
 https://bugzilla.mozilla.org/show_bug.cgi?id=695438#c25
Andre Bargull pointed out that the current definition of what is an indexed property of a typed array in the spec is still pending changes:
 https://bugs.ecmascript.org/show_bug.cgi?id=2049

I think the intended changes sound good, but I just wanted to double check:
- negative integers will unconditionally be undefined (this one is critical to maintain)
- super-large integers will now also be unconditionally undefined (instead of own properties)

Less, that's correct.  Actually any integer >= the length of the typed array unconditionally yields undefined.

There is still a remaining compatibility issue we should probably consider.

It would be nice if using +/-Infinity as an index also yielded undefined.

However, assuming that legacy TypedArray implementations applied the ToUint32 conversion to the index, infinities would be equivalent to 0 as an index.

I can easily imagine that there are buggy programs whose behavior depend upon that behavior and hence we probably need to preserve that behavior.

What do you think? What happens with infinities in your fast paths?

Allen
```

**Comment by leobalter** [comment]
> In your table for: new Int8Array( new ArrayBuffer(128), 0, x ).length
you say the spec. produces 2**53-1 when x is +∞. My reading of both the ES2015 spec. and the current draft is that RangeError should occur. See Step 14.c.
Similar issue for the DateView table. See Step 11.b.
What are you interpreting differently from me?

That's my bad on merging two different tables I've had made. The `new Int8Array( new ArrayBuffer(128), 0, x ).length` reflects only the tests for the runtimes. For the specs + pr, the results I presented are expected on the length conversion. I should have considered [Step 14.c](http://tc39.github.io/ecma262/#sec-typedarray-buffer-byteoffset-length) but I missed that ignoring anything beyond 14.a for the TypedArray and 11.a for the DataView.

On this case, my data is wrong, but the RangeError would be preserved on the spec.

**Comment by leobalter** [comment]
> Regarding the Khronos spec. I think the original author (Ken) just used WebIDL conventions to describe the signatures To understand them you have to look at the Khronos signatures and then look at the conversion rules for the WebIDL types used in the WebIDL spec that was contemporary with the Khronos spec. For example, x in the signature applicable to the above table is specified to be "unsigned long" so the this conversion applies to it.

That's very helpful, I cay update the table with this info if it helps, but I believe these changes are close to that spec, with a few minor differences. 

e.g.:

- similar behaviors:
  - non-integer numbers are cast as integers;
  - throwing for negative numbers, Infinity or excessive numbers
- different behaviors:
  - Throwing RangeErrors instead of TypeErrors
  - Casting NaN to 0 instead of throwing a RangeError (I'm not sure about this one, I'm slightly inclined to change it to throw a RangeError too, but I'm following implemenations mostly)
  - The max index is `2 ** 53 - 1` instead of `2 ** 32 - 1`.

> infinities would be equivalent to 0 as an index

We are not doing this on the constructor, although a few implementations are casting Infinity to 0 as you can notice from the table. That's not reflected on the specs.

**Comment by zloirock** [comment]
@leobalter sorry, I missed your comment. Sure, conversion floats to integers without error for `TypedArray` should prevent this problem.

**Comment by leobalter** [comment]
Thanks, @zloirock!

@allenwb I added notes on https://github.com/tc39/ecma262/pull/410#issuecomment-208589878 to mention the Infinity values may cause the RangeError.

I also believe that might not trigger the RangeError all the time, assuming `offset + newByteLength > bufferByteLength` can be false as in: `0 + (2 ** 53 - 1) > (2 ** 53 - 1)`.

**Comment by leobalter** [comment]
This reached consensus in the last meeting of May 2016. 

@bterlson can we merge this? 👍
